x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/slowlog/EsqlSlowLog.java

```diff
@@ -0,0 +1,160 @@
+    public void setQueryWarnThreshold(TimeValue queryWarnThreshold) {
+        this.queryWarnThreshold = queryWarnThreshold.nanos();
+    }
+
+    public void setQueryInfoThreshold(TimeValue queryInfoThreshold) {
+        this.queryInfoThreshold = queryInfoThreshold.nanos();
+    }
+
+    public void setQueryDebugThreshold(TimeValue queryDebugThreshold) {
+        this.queryDebugThreshold = queryDebugThreshold.nanos();
+    }
+
+    public void setQueryTraceThreshold(TimeValue queryTraceThreshold) {
+        this.queryTraceThreshold = queryTraceThreshold.nanos();
+    }
+
+    public void setIncludeUser(boolean includeUser) {
+        this.includeUser = includeUser;
+    }
+
+    static final class Message {
+
+        private static String escapeJson(String text) {
+            byte[] sourceEscaped = JsonStringEncoder.getInstance().quoteAsUTF8(text);
+            return new String(sourceEscaped, StandardCharsets.UTF_8);
+        }
+
+        public static ESLogMessage of(Result esqlResult, String query, User user) {
+            Map<String, Object> jsonFields = prepareMap(esqlResult, query, true, user);
+
+            return new ESLogMessage().withFields(jsonFields);
+        }
+
+        public static ESLogMessage of(String query, long took, Exception exception, User user) {
+            Map<String, Object> jsonFields = prepareMap(null, query, false, user);
+            jsonFields.put("elasticsearch.slowlog.error.message", exception.getMessage() == null ? "" : exception.getMessage());
+            jsonFields.put("elasticsearch.slowlog.error.type", exception.getClass().getName());
+            jsonFields.put("elasticsearch.slowlog.took", took);
+            jsonFields.put("elasticsearch.slowlog.took_millis", took / 1_000_000);
+            return new ESLogMessage().withFields(jsonFields);
+        }
+
+        private static Map<String, Object> prepareMap(Result esqlResult, String query, boolean success, User user) {
+            Map<String, Object> messageFields = new HashMap<>();
+            if (user != null) {
+                messageFields.put("user.name", user.principal());
+            }
+            if (esqlResult != null) {
+                messageFields.put("elasticsearch.slowlog.took", esqlResult.executionInfo().overallTook().nanos());
+                messageFields.put("elasticsearch.slowlog.took_millis", esqlResult.executionInfo().overallTook().millis());
+                messageFields.put("elasticsearch.slowlog.planning.took", esqlResult.executionInfo().planningTookTime().nanos());
+                messageFields.put("elasticsearch.slowlog.planning.took_millis", esqlResult.executionInfo().planningTookTime().millis());
+            }
+            String source = escapeJson(query);
+            messageFields.put("elasticsearch.slowlog.success", success);
+            messageFields.put("elasticsearch.slowlog.search_type", "ESQL");
+            messageFields.put("elasticsearch.slowlog.query", source);
+            return messageFields;
```

Comment:
Extract constants out of these repetitive strings including during their construction, such as the common prefix ("elasticsearch.slowlog")

Also there's a repetition on took/took_millis, which seems to be overridden always in of method.